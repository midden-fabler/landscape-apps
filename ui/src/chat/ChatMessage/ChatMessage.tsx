/* eslint-disable react/no-unused-prop-types */
import React, { useCallback, useEffect, useRef, useState } from 'react';
import cn from 'classnames';
import _ from 'lodash';
import debounce from 'lodash/debounce';
import f from 'lodash/fp';
import { BigInteger } from 'big-integer';
import { daToUnix } from '@urbit/api';
import { format, formatDistanceToNow, formatRelative, isToday } from 'date-fns';
import { NavLink, useParams } from 'react-router-dom';
import { useInView } from 'react-intersection-observer';
import { ChatBrief, ChatStory, ChatWrit } from '@/types/chat';
import Author from '@/chat/ChatMessage/Author';
// eslint-disable-next-line import/no-cycle
import ChatContent from '@/chat/ChatContent/ChatContent';
import ChatReactions from '@/chat/ChatReactions/ChatReactions';
import DateDivider from '@/chat/ChatMessage/DateDivider';
import ChatMessageOptions from '@/chat/ChatMessage/ChatMessageOptions';
import {
  usePact,
  useChatState,
  useIsMessageDelivered,
  useIsMessagePosted,
  useWrit,
  useMessageToggler,
} from '@/state/chat';
import Avatar from '@/components/Avatar';
import DoubleCaretRightIcon from '@/components/icons/DoubleCaretRightIcon';
import UnreadIndicator from '@/components/Sidebar/UnreadIndicator';
import { whomIsDm, whomIsMultiDm } from '@/logic/utils';
import { useIsMobile } from '@/logic/useMedia';
import useLongPress from '@/logic/useLongPress';
import {
  useChatDialog,
  useChatHovering,
  useChatInfo,
  useChatStore,
} from '../useChatStore';
import ReactionDetails from '../ChatReactions/ReactionDetails';

export interface ChatMessageProps {
  whom: string;
  time: BigInteger;
  writ: ChatWrit;
  newAuthor?: boolean;
  newDay?: boolean;
  hideReplies?: boolean;
  hideOptions?: boolean;
  isLast?: boolean;
  isLinked?: boolean;
  isScrolling?: boolean;
}

function briefMatches(brief: ChatBrief, id: string): boolean {
  return brief['read-id'] === id;
}

const mergeRefs =
  (...refs: any[]) =>
  (node: any) => {
    refs.forEach((ref) => {
      if (!ref) {
        return;
      }

      /* eslint-disable-next-line no-param-reassign */
      ref.current = node;
    });
  };

const hiddenMessage: ChatStory = {
  block: [],
  inline: [
    {
      italics: [
        'You have hidden this message. You can unhide it from the options menu.',
      ],
    },
  ],
};

const ChatMessage = React.memo<
  ChatMessageProps & React.RefAttributes<HTMLDivElement>
>(
  React.forwardRef<HTMLDivElement, ChatMessageProps>(
    (
      {
        whom,
        time,
        writ,
        newAuthor = false,
        newDay = false,
        hideReplies = false,
        hideOptions = false,
        isLast = false,
        isLinked = false,
        isScrolling = false,
      }: ChatMessageProps,
      ref
    ) => {
      const { seal, memo } = writ;
      const container = useRef<HTMLDivElement>(null);
      const { idShip, idTime } = useParams<{
        idShip: string;
        idTime: string;
      }>();
      const isThread = !!idShip && !!idTime;
      const threadOpId = isThread ? `${idShip}/${idTime}` : '';
      const isThreadOp = threadOpId === seal.id && hideReplies;
      const isMobile = useIsMobile();
      const isThreadOnMobile = isThread && isMobile;
      const chatInfo = useChatInfo(whom);
      const unread = chatInfo?.unread;
      const unreadId = unread?.brief['read-id'];
      const { hovering, setHovering } = useChatHovering(whom, writ.seal.id);
      const { isHidden } = useMessageToggler(writ.seal.id);
      const { open: pickerOpen } = useChatDialog(whom, writ.seal.id, 'picker');
      const { ref: viewRef } = useInView({
        threshold: 1,
        onChange: useCallback(
          (inView: boolean) => {
            // if no tracked unread we don't need to take any action
            if (!unread) {
              return;
            }

            const { brief, seen } = unread;
            /* the first fire of this function
               which we don't to do anything with. */
            if (!inView && !seen) {
              return;
            }

            const {
              seen: markSeen,
              read,
              delayedRead,
            } = useChatStore.getState();
            const { markRead } = useChatState.getState();

            /* once the unseen marker comes into view we need to mark it
               as seen and start a timer to mark it read so it goes away.
               we ensure that the brief matches and hasn't changed before
               doing so. we don't want to accidentally clear unreads when
               the state has changed
            */
            if (inView && briefMatches(brief, writ.seal.id) && !seen) {
              markSeen(whom);
              delayedRead(whom, () => markRead(whom));
              return;
            }

            /* finally, if the marker transitions back to not being visible,
              we can assume the user is done and clear the unread. */
            if (!inView && unread && seen) {
              read(whom);
              markRead(whom);
            }
          },
          [unread, whom, writ.seal.id]
        ),
      });
      const isMessageDelivered = useIsMessageDelivered(seal.id);
      const isMessagePosted = useIsMessagePosted(seal.id);
      const isReplyOp = chatInfo?.replying === writ.seal.id;

      const unix = new Date(daToUnix(time));

      const pact = usePact(whom);

      const numReplies = seal.replied.length;
      const repliesContainsUnreadId = unreadId
        ? seal.replied.includes(unreadId)
        : false;
      const replyAuthors = _.flow(
        f.map((k: string) => {
          const t = pact?.index[k];
          const mess = t ? pact.writs.get(t) : undefined;
          if (!mess) {
            return undefined;
          }
          return mess.memo.author;
        }),
        f.compact,
        f.uniq,
        f.take(3)
      )(seal.replied);
      const repliesSortedByTime = seal.replied.sort((a, b) => {
        const aTime = useChatState.getState().getTime(whom, a);
        const bTime = useChatState.getState().getTime(whom, b);

        return aTime.compare(bTime);
      });
      const lastReply = _.last(repliesSortedByTime);
      const { entry: lastReplyWrit } = useWrit(whom, lastReply ?? '');
      const lastReplyTime = lastReplyWrit
        ? new Date(daToUnix(lastReplyWrit[0]))
        : new Date();

      const hover = useRef(false);
      const setHover = useRef(
        debounce(() => {
          if (hover.current) {
            setHovering(true);
          }
        }, 100)
      );
      const onOver = useCallback(() => {
        // If we're already hovering, don't do anything
        // If we're the thread op and this isn't on mobile, don't do anything
        // This is necessary to prevent the hover from appearing
        // in the thread when the user hovers in the main scroll window.
        if (hover.current === false && (!isThreadOp || isThreadOnMobile)) {
          hover.current = true;
          setHover.current();
        }
      }, [isThreadOp, isThreadOnMobile]);
      const onOut = useRef(
        debounce(
          () => {
            hover.current = false;
            setHovering(false);
          },
          50,
          { leading: true }
        )
      );

      const [optionsOpen, setOptionsOpen] = useState(false);
      const [reactionDetailsOpen, setReactionDetailsOpen] = useState(false);
      const { action, actionId, handlers } = useLongPress({ withId: true });

      const handleReactionDetailsOpened = useCallback(() => {
        setReactionDetailsOpen(true);
      }, []);

      useEffect(() => {
        if (!isMobile) {
          return;
        }

        if (action === 'longpress') {
          if (actionId === 'reactions-target') {
            setReactionDetailsOpen(true);
          } else {
            setOptionsOpen(true);
          }
        }
      }, [action, actionId, isMobile]);

      useEffect(() => {
        if (isMobile) {
          return;
        }

        // If we're the thread op, don't show options.
        // Options are shown for the threadOp in the main scroll window.
        setOptionsOpen(
          (hovering || pickerOpen) &&
            !hideOptions &&
            !isScrolling &&
            !isThreadOp
        );
      }, [
        isMobile,
        hovering,
        pickerOpen,
        hideOptions,
        isScrolling,
        isThreadOp,
      ]);

      return (
        <div
          ref={mergeRefs(ref, container)}
          className={cn('flex flex-col break-words', {
            'pt-3': newAuthor,
            'pb-2': isLast,
          })}
          onMouseEnter={onOver}
          onMouseLeave={onOut.current}
          data-testid="chat-message"
          id="chat-message-target"
          {...handlers}
        >
          {unread && briefMatches(unread.brief, writ.seal.id) ? (
            <DateDivider
              date={unix}
              unreadCount={unread.brief.count}
              ref={viewRef}
            />
          ) : null}
          {newDay ? <DateDivider date={unix} /> : null}
          {newAuthor ? (
            <Author ship={memo.author} date={unix} hideRoles={isThread} />
          ) : null}
          <div className="group-one relative z-0 flex w-full select-none sm:select-auto">
            <ChatMessageOptions
              open={optionsOpen}
              onOpenChange={setOptionsOpen}
              hideThreadReply={hideReplies}
              whom={whom}
              writ={writ}
              hideReply={whomIsDm(whom) || whomIsMultiDm(whom) || hideReplies}
              openReactionDetails={handleReactionDetailsOpened}
            />
            <div className="-ml-1 mr-1 py-2 text-xs font-semibold text-gray-400 opacity-0 sm:group-one-hover:opacity-100">
              {format(unix, 'HH:mm')}
            </div>
            <div className="wrap-anywhere flex w-full">
              <div
                className={cn(
                  'flex w-full min-w-0 grow flex-col space-y-2 rounded py-1 pl-3 pr-2 sm:group-one-hover:bg-gray-50',
                  isReplyOp && 'bg-gray-50',
                  !isMessageDelivered && !isMessagePosted && 'text-gray-400',
                  isLinked && 'bg-blue-softer'
                )}
              >
                {isHidden ? (
                  <ChatContent
                    story={hiddenMessage}
                    isScrolling={isScrolling}
                    writId={seal.id}
                  />
                ) : 'story' in memo.content ? (
                  <ChatContent
                    story={memo.content.story}
                    isScrolling={isScrolling}
                    writId={seal.id}
                  />
                ) : null}
                {Object.keys(seal.feels).length > 0 && (
                  <>
                    <ChatReactions
                      id="reactions-target"
                      seal={seal}
                      whom={whom}
                    />
                    <ReactionDetails
                      open={reactionDetailsOpen}
                      onOpenChange={setReactionDetailsOpen}
                      feels={seal.feels}
                    />
                  </>
                )}
                {numReplies > 0 && !hideReplies ? (
                  <NavLink
                    to={`message/${seal.id}`}
                    className={({ isActive }) =>
                      cn(
                        'default-focus group -ml-2 whitespace-nowrap rounded p-2 text-sm font-semibold text-gray-800',
                        isActive
                          ? 'is-active bg-gray-50 [&>div>div>.reply-avatar]:outline-gray-50'
                          : '',
                        isLinked
                          ? '[&>div>div>.reply-avatar]:outline-blue-100 dark:[&>div>div>.reply-avatar]:outline-blue-900'
                          : ''
                      )
                    }
                  >
                    <div className="flex items-center">
                      <div className="mr-2 flex flex-row-reverse">
                        {replyAuthors.map((ship, i) => (
                          <div
                            key={ship}
                            className={cn(
                              'reply-avatar relative h-6 w-6 rounded bg-white outline outline-2 outline-white sm:group-one-focus-within:outline-gray-50 sm:group-one-hover:outline-gray-50',
                              i !== 0 && '-mr-3'
                            )}
                          >
                            <Avatar key={ship} ship={ship} size="xs" />
                          </div>
                        ))}
                      </div>

                      <span
                        className={cn(
                          repliesContainsUnreadId ? 'text-blue' : 'mr-2'
                        )}
                      >
                        {numReplies} {numReplies > 1 ? 'replies' : 'reply'}{' '}
                      </span>
                      {repliesContainsUnreadId ? (
                        <UnreadIndicator
                          className="h-6 w-6 text-blue transition-opacity"
                          aria-label="Unread replies in this thread"
                        />
                      ) : null}
                      <span className="text-gray-400">
                        <span className="hidden sm:inline-block">
                          Last reply&nbsp;
                        </span>
                        <span className="inline-block first-letter:capitalize sm:first-letter:normal-case">
                          {isToday(lastReplyTime)
                            ? `${formatDistanceToNow(lastReplyTime)} ago`
                            : formatRelative(lastReplyTime, new Date())}
                        </span>
                      </span>
                    </div>
                  </NavLink>
                ) : null}
              </div>
              <div className="relative flex w-5 items-end rounded-r sm:group-one-hover:bg-gray-50">
                {!isMessageDelivered && (
                  <DoubleCaretRightIcon
                    className="absolute left-0 bottom-2 h-5 w-5"
                    primary={isMessagePosted ? 'text-black' : 'text-gray-200'}
                    secondary="text-gray-200"
                  />
                )}
              </div>
            </div>
          </div>
        </div>
      );
    }
  )
);

export default ChatMessage;
