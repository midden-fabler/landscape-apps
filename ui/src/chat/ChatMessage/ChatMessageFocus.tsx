import { useMemo, useState } from 'react';
import { useLocation, useParams } from 'react-router';
import cn from 'classnames';
import { Transition, motion } from 'framer-motion';
import { ChatWrit } from '@/types/chat';
import { useDismissNavigate } from '@/logic/routing';
import ElipsisIcon from '@/components/icons/EllipsisIcon';
import ConfirmationModal from '@/components/ConfirmationModal';
import Author from './Author';
import ChatContent from '../ChatContent/ChatContent';
import ReactionDetails from '../ChatReactions/ReactionDetails';
import useChatMessage, { ChatMessage } from './useChatMessageActions';
import ReactionSheet from '../ChatReactions/EmojiPickerSheet';

interface ChatOptionsMenuProps {
  chatMessage: ChatMessage;
  showAllReactions: () => void;
  initiateDelete: () => void;
}
function ChatOptionsMenu({
  chatMessage,
  showAllReactions,
  initiateDelete,
}: ChatOptionsMenuProps) {
  const dismiss = useDismissNavigate();

  const handleCopy = () => {
    chatMessage.actions.copy();
    setTimeout(() => {
      dismiss();
    }, 1000);
  };

  const handleHideToggle = () => {
    if (chatMessage.state.isHidden) {
      chatMessage.actions.show();
    } else {
      chatMessage.actions.hide();
    }
    dismiss();
  };

  return (
    <motion.div
      initial={{ opacity: 0, x: 0 }}
      animate={{ opacity: 1, x: 0 }}
      transition={{ duration: 0.4 }}
      onClick={(e) => e.stopPropagation()}
      className="flex w-48 flex-col rounded-lg border border-gray-200 bg-gray-100 text-lg"
    >
      <button
        className="border-b border-gray-200 py-3"
        onClick={showAllReactions}
      >
        View Reactions
      </button>
      {chatMessage.context.canReply && (
        <button
          className="border-b border-gray-200 py-3"
          onClick={chatMessage.actions.reply}
        >
          Reply
        </button>
      )}
      {!chatMessage.context.inThread && (
        <button
          className="border-b border-gray-200 py-3"
          onClick={chatMessage.actions.goToThread}
        >
          Start Thread
        </button>
      )}
      {chatMessage.context.canCopy && (
        <button className="border-b border-gray-200 py-3" onClick={handleCopy}>
          {chatMessage.state.didCopy ? 'Copied!' : 'Copy Link'}
        </button>
      )}
      {chatMessage.context.canHide && (
        <button
          className="border-b border-gray-200 py-3"
          onClick={handleHideToggle}
        >
          {chatMessage.state.isHidden ? 'Show Message' : 'Hide Message'}
        </button>
      )}
      {chatMessage.context.canDelete && (
        <button className="py-3 text-red" onClick={initiateDelete}>
          Delete
        </button>
      )}
    </motion.div>
  );
}

function QuickReactOption({
  shortcode,
  addFeel,
  transition,
}: {
  shortcode: string;
  addFeel: (code: string) => void;
  transition: Transition;
}) {
  return (
    <motion.div
      initial={{ opacity: 0 }}
      animate={{ opacity: 1 }}
      transition={transition}
      className="flex items-center justify-center p-2"
      onClick={() => addFeel(shortcode)}
    >
      <em-emoji shortcodes={shortcode} size="1.6em" />
    </motion.div>
  );
}

function EmojiQuickReact({
  addFeel,
  openPicker,
}: {
  addFeel: (shortcode: string) => void;
  openPicker: () => void;
}) {
  return (
    <motion.div
      initial={{ scale: 0, x: 0 }}
      animate={{ scale: 1, x: 0 }}
      transition={{ type: 'spring', stiffness: 200, duration: 0.005 }}
      className="flex w-[280px] items-center justify-evenly rounded-[24px] bg-white py-2 px-3"
      onClick={(e) => e.stopPropagation()}
    >
      <QuickReactOption
        transition={{
          delay: 0.05,
          type: 'inertia',
          velocity: 300,
          power: 0.9,
          timeConstant: 150,
          restDelta: 1,
        }}
        shortcode=":+1:"
        addFeel={addFeel}
      />
      <QuickReactOption
        transition={{
          delay: 0.1,
          type: 'inertia',
          velocity: 300,
          power: 0.9,
          timeConstant: 150,
          restDelta: 1,
        }}
        shortcode=":heart:"
        addFeel={addFeel}
      />
      <QuickReactOption
        transition={{
          delay: 0.15,
          type: 'inertia',
          velocity: 300,
          power: 0.9,
          timeConstant: 150,
          restDelta: 1,
        }}
        shortcode=":cyclone:"
        addFeel={addFeel}
      />
      <QuickReactOption
        transition={{
          delay: 0.2,
          type: 'inertia',
          velocity: 300,
          power: 0.9,
          timeConstant: 150,
          restDelta: 1,
        }}
        shortcode=":seedling:"
        addFeel={addFeel}
      />
      <motion.div
        initial={{ opacity: 0 }}
        animate={{ opacity: 1 }}
        transition={{
          delay: 0.25,
          type: 'inertia',
          velocity: 300,
          power: 0.9,
          timeConstant: 150,
          restDelta: 1,
        }}
        className="flex h-10 w-10 items-center justify-center rounded-full bg-gray-100 p-3"
        onClick={openPicker}
      >
        <ElipsisIcon className="h-8 w-8" />
      </motion.div>
    </motion.div>
  );
}

function getScaleFactor(targetHeight: number, height?: number) {
  let scale = 1;
  if (height) {
    if (height > targetHeight) {
      scale = targetHeight / height;
    }
  }
  return scale;
}

export default function ChatMessageFocus() {
  const params = useParams<{
    ship?: string;
    name?: string;
    chShip?: string;
    chName?: string;
    idShip?: string;
    idTime?: string;
  }>();
  const location = useLocation();
  const dismiss = useDismissNavigate();

  const [showPicker, setShowPicker] = useState(false);
  const [showAllReactions, setShowAllReactions] = useState(false);
  const [showDeleteConfirm, setShowDeleteConfirm] = useState(false);

  const showOptions = !showPicker && !showAllReactions && !showDeleteConfirm;
  const showMessage = !showDeleteConfirm;

  const { writ, height } = location.state as {
    writ: ChatWrit;
    height?: number;
  };
  const { memo, seal } = writ;
  const chatMessage = useChatMessage(writ);

  // we need to constrain the height of the focused message, 40% of window height is
  // arbitrarily chosen
  const messageTargetHeight = Math.round(window.innerHeight * 0.3);
  const scaleFactor = useMemo(
    () => getScaleFactor(messageTargetHeight, height),
    [height, messageTargetHeight]
  );

  // if we scale the message, the layout doesn't update automatically, so we
  // need to measure and position manually
  const quickReactionEl = document.getElementById('quickReactionMenu');
  const messageContainerEl = document.getElementById('focusedMessageContainer');
  const authorEl = document.getElementById('author');
  const messageContentEl = document.getElementById('focusedMessageContent');
  const optionsMenuEl = document.getElementById('chatOptionsMenu');

  let messageContainerHeight = '0px';
  let messageContentTop = '0px';
  if (authorEl && messageContentEl && messageContainerEl && scaleFactor !== 1) {
    const contentRec = messageContentEl!.getBoundingClientRect();
    messageContainerHeight = `${Math.round(
      contentRec.bottom - contentRec.top + authorEl.clientHeight
    )}px`;
    messageContentTop = `${authorEl.clientHeight}px`;
  }

  const focusedMessageMargin = 14;
  let quickReactTop = '0px';
  let quickReactLeft = '18px';
  if (messageContainerEl && quickReactionEl) {
    const fmRect = messageContainerEl!.getBoundingClientRect();
    quickReactTop = `${Math.round(
      fmRect.top - quickReactionEl.clientHeight - focusedMessageMargin
    )}px`;
    quickReactLeft = `${fmRect.left}px`;
  }

  const optionsMenuTopMargin = 14;
  let optionsMenuTop = '0px';
  let optionsMenuLeft = '0px';
  if (messageContainerEl && optionsMenuEl) {
    const fmRect = messageContainerEl!.getBoundingClientRect();
    optionsMenuTop = `${Math.round(fmRect.bottom + optionsMenuTopMargin)}px`;
    optionsMenuLeft = `${Math.round(fmRect.right) - 192}px`;
  }

  const overlayClick = (
    event: React.MouseEvent<HTMLDivElement, MouseEvent>
  ) => {
    dismiss();
  };

  const containerClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    dismiss();
  };

  const addFeel = async (shortcode: string) => {
    await chatMessage.actions.react(shortcode);
    dismiss();
  };

  const delMessage = async () => {
    try {
      await chatMessage.actions.del();
    } finally {
      dismiss();
    }
  };

  if ('story' in memo.content) {
    return (
      <>
        <div
          id="overlay"
          className="fixed z-30 h-full w-full bg-black/20 backdrop-blur-md"
          onClick={overlayClick}
        />

        <div
          className="z-40 h-full w-full select-none"
          onClick={containerClick}
        >
          <div
            id="quickReactionMenu"
            className={
              !showOptions || !messageContainerEl ? 'invisible' : 'absolute'
            }
            style={{ top: quickReactTop, left: quickReactLeft }}
            onClick={(e) => e.stopPropagation()}
          >
            <EmojiQuickReact
              addFeel={addFeel}
              openPicker={() => setShowPicker(true)}
            />
          </div>

          <motion.div
            initial={{ opacity: 0, x: 0 }}
            animate={{
              opacity: 1,
              x: 0,
            }}
            exit={{ opacity: 0 }}
            transition={{ duration: 0.05 }}
            layoutId={seal.id}
            onClick={(e) => e.stopPropagation()}
          >
            {scaleFactor === 1 ? (
              <div className="absolute top-[35%] flex w-full justify-start">
                <div
                  id="focusedMessageContainer"
                  className="ml-6 flex min-w-[280px] max-w-[80%] flex-col items-start justify-center rounded-xl bg-white p-4"
                >
                  <Author className="mb-2" ship={memo.author} hideRoles />
                  <div id="focusedMessageContent">
                    <ChatContent
                      story={memo.content.story}
                      writId={writ.seal.id}
                      className="select-auto"
                    />
                  </div>
                </div>
              </div>
            ) : (
              <div
                id="focusedMessageContainer"
                className={cn(
                  'absolute top-[20%] ml-6 flex w-[300px] flex-col rounded-xl bg-white'
                )}
                style={{ height: messageContainerHeight }}
                onClick={(e) => e.stopPropagation()}
              >
                <div
                  id="author"
                  className="flex items-center justify-start pt-4 pl-4"
                >
                  <Author className="" ship={memo.author} hideRoles />
                </div>
                <div className="flex flex-1 items-center justify-center overflow-hidden">
                  <div
                    id="focusedMessageContent"
                    className="pt-3 pb-4"
                    style={{ transform: `scale(${scaleFactor})` }}
                  >
                    <ChatContent
                      story={memo.content.story}
                      writId={writ.seal.id}
                      className="select-auto"
                    />
                  </div>
                </div>
              </div>
            )}
          </motion.div>

          <div
            id="chatOptionsMenu"
            className={
              !showOptions || !messageContainerEl ? 'invisible' : 'absolute'
            }
            style={{ top: optionsMenuTop, left: optionsMenuLeft }}
            onClick={(e) => e.stopPropagation()}
          >
            <ChatOptionsMenu
              chatMessage={chatMessage}
              showAllReactions={() => setShowAllReactions(true)}
              initiateDelete={() => setShowDeleteConfirm(true)}
            />
          </div>

          {showAllReactions && (
            <div>
              <ReactionDetails
                open={showAllReactions}
                onOpenChange={setShowAllReactions}
                feels={writ.seal.feels}
              />
            </div>
          )}

          {showPicker && (
            <ReactionSheet
              onEmojiSelect={addFeel}
              open={showPicker}
              onOpenChange={setShowPicker}
            />
          )}

          {showDeleteConfirm && (
            <ConfirmationModal
              title="Delete Message"
              message="Are you sure you want to delete this message?"
              onConfirm={delMessage}
              open={showDeleteConfirm}
              setOpen={setShowDeleteConfirm}
              confirmText="Delete"
              loading={chatMessage.state.isDeleting}
            />
          )}
        </div>
      </>
    );
  }

  return null;
}
