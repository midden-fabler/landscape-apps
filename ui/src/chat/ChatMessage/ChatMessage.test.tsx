import { describe, it, expect, beforeEach } from 'vitest';
import React from 'react';
import { unixToDa } from '@urbit/api';
import { TooltipProvider } from '@radix-ui/react-tooltip';
import ChatMessage from './ChatMessage';
import { makeFakeChatWrit } from '../../mocks/chat';
import { render } from '../../../test/utils';

describe('ChatMessage', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });
  it('renders as expected', () => {
    const date = new Date(2021, 1, 1, 13);
    const writ = makeFakeChatWrit(
      1,
      '~finned-palmer',
      {
        block: [],
        inline: [{ bold: ['A bold test message'] }, 'with some more text'],
      },
      undefined
    );
    const da = unixToDa(date.valueOf());
    const { asFragment } = render(
      <TooltipProvider>
        <ChatMessage time={da} whom="~zod/test" writ={writ} newAuthor newDay />
      </TooltipProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
