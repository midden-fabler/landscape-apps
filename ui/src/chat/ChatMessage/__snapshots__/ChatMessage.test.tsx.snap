// Vitest Snapshot v1

exports[`ChatMessage > renders as expected 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col break-words pt-3"
    data-testid="chat-message"
  >
    <div
      class="flex w-full items-center pt-4 pb-6"
    >
      <div
        class="h-[2px] w-8 rounded-sm bg-gray-200"
      >
         
      </div>
      <span
        class="whitespace-nowrap px-3 font-semibold text-gray-400"
      >
        February 1st
      </span>
      <div
        class="h-[2px] w-full rounded-sm bg-gray-200"
      >
         
      </div>
    </div>
    <div
      class="group-two align-center flex items-center py-1 space-x-3"
    >
      <div
        class="shrink-0"
      >
        <div
          class="relative flex flex-none items-center justify-center rounded bg-black w-6 h-6 rounded p-1.5 cursor-pointer"
          style="background-color: rgb(0, 0, 0);"
        >
          <svg
            class=""
            style="display: block; width: 12px; height: 12px;"
            version="1.1"
            viewBox="0 0 12 12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="#000000"
              height="12px"
              width="12px"
              x="0px"
              y="0px"
            />
            <g
              transform="matrix(0.046875,0,0,0.046875,0,0)"
            >
              <g
                fill="none"
                transform="matrix(1,0,0,1,0,0)"
              >
                <rect
                  dataisgeon="true"
                  fill="white"
                  height="128"
                  stroke="#000000"
                  stroke-linecap="square"
                  stroke-width="0.8px"
                  vector-effect="non-scaling-stroke"
                  width="128"
                />
              </g>
              <g
                fill="none"
                transform="matrix(1,0,0,1,128,0)"
              >
                <circle
                  cx="64"
                  cy="64"
                  dataisgeon="true"
                  fill="white"
                  r="64"
                  stroke="#000000"
                  stroke-linecap="square"
                  stroke-width="0.8px"
                  vector-effect="non-scaling-stroke"
                />
              </g>
              <g
                fill="none"
                transform="matrix(1,0,0,1,0,128)"
              >
                <path
                  d="M0.0541 0C70.7217 0.0292317 128 57.3256 128 128C57.3177 128 0.0164917 70.7089 7.62806e-06 0.0305091C7.62851e-06 0.0203397 -4.44317e-10 0.01017 0 0H0.0541Z"
                  dataisgeon="true"
                  fill="white"
                  stroke="#000000"
                  stroke-linecap="square"
                  stroke-width="0.8px"
                  vector-effect="non-scaling-stroke"
                />
              </g>
              <g
                fill="none"
                transform="matrix(1,0,0,1,128,128)"
              >
                <path
                  d="M64 128H0L5.59506e-06 0L64 5.59506e-06C99.3462 8.68512e-06 128 28.6538 128 64C128 99.3462 99.3462 128 64 128Z"
                  dataisgeon="true"
                  fill="white"
                  stroke="#000000"
                  stroke-linecap="square"
                  stroke-width="0.8px"
                  vector-effect="non-scaling-stroke"
                />
              </g>
            </g>
          </svg>
        </div>
      </div>
      <div
        class="text-md shrink cursor-pointer font-semibold"
      >
        <span
          class="text-md break-all font-semibold leading-6 line-clamp-1"
          title=""
        >
          <span
            aria-hidden="true"
          >
            ~
          </span>
          <span>
            finned
          </span>
          <span
            aria-hidden="true"
          >
            -
          </span>
          <span
            aria-hidden="false"
          >
            palmer
          </span>
        </span>
      </div>
      <span
        class="-mb-0.5 hidden shrink-0 text-sm font-semibold text-gray-500 group-two-hover:block"
      >
        <time
          datetime="2021-02-01T13:00:00.000Z"
        >
          February 1st, 2021 
          <span
            aria-hidden="true"
          >
            •
          </span>
           13:00
        </time>
      </span>
      <span
        class="-mb-0.5 block shrink-0 text-sm font-semibold text-gray-500 group-two-hover:hidden"
      >
        February 1st 
        <span
          aria-hidden="true"
        >
          •
        </span>
         
        <time
          datetime="13:00"
        >
          13:00
        </time>
      </span>
    </div>
    <div
      class="group-one relative z-0 flex w-full"
    >
      <div
        class="-ml-1 mr-1 py-2 text-xs font-semibold text-gray-400 opacity-0 group-one-hover:opacity-100"
      >
        13:00
      </div>
      <div
        class="wrap-anywhere flex w-full"
      >
        <div
          class="flex w-full grow flex-col space-y-2 rounded py-1 pl-3 pr-2 group-one-hover:bg-gray-50"
        >
          <div
            class="leading-6"
          >
            <strong>
              <span>
                A bold test message
              </span>
            </strong>
            <span>
              with some more text
            </span>
          </div>
        </div>
        <div
          class="relative flex w-5 items-end rounded-r group-one-hover:bg-gray-50"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
